### ISSUE TITLE
[Bug]: go 1.20.6: "http: invalid Host header, host port waiting failed"

### ISSUE THREAD

**Issue by srabraham** [testcontainers/testcontainers-go#1359]
### Testcontainers version

0.21.0

### Using the latest Testcontainers version?

Yes

### Host OS

OSX and Linux

### Host arch

OSX arm64, Linux amd64

### Go version

1.20.6

### What happened?

With Go 1.20.6, I start getting an error "http: invalid Host header, host port waiting failed" when setting up new containers. From doing some debugging, it seems to be coming from the ryuk setup. This appears to be the relevant change in the standard library: https://github.com/golang/go/issues/61076, since it introduces that check that returns "invalid Host header". From doing the debugging in the screenshot below, the Host that Go is choking on is "/var/run/docker.sock".

![image](https://github.com/testcontainers/testcontainers-go/assets/23407555/444eafae-e6da-43ea-81ba-3d2cf36793a6)
![image](https://github.com/testcontainers/testcontainers-go/assets/23407555/5cd40cb6-09a7-4e83-b576-62a0d6e6efe7)
![image](https://github.com/testcontainers/testcontainers-go/assets/23407555/035c7215-d80b-40f2-9001-3f8917b7a0d2)

**Comment by Akaame** [comment]
```golang
assert.True(t, httpguts.ValidHostHeader("/var/run/docker.sock"))
```

fails because `/` is not a valid char in a host name.

**Comment by srabraham** [comment]
@Akaame yes, I agree with you on that. The problem is that in my company's codebase, we're not setting that value ("/var/run/docker.sock") anywhere, so I'm thinking it's being picked up in testcontainers code. For example, reaper.go seems to set that value by default as of not too long ago. See https://github.com/testcontainers/testcontainers-go/pull/1161

**Comment by dicrescenzoa** [comment]
I've had the same problem and I've realized that this is happening with golang 1.20.6. 
Downgrading Golang to 1.20.3 resolved the problem.

**Comment by hamin** [comment]
I've been getting this same problem since yesterday out of the blue. Is there any other way to resolve this without downgrading the Go version?

**Comment by kshitij10496** [comment]
I also faced this issue and resolved it temporarily by downgrading to Go 1.20.5

**Comment by srabraham** [comment]
As I mentioned in the original post, this issue is introduced in Go 1.20.6, via https://github.com/golang/go/issues/61076

@Slijkhuis: thanks! I'll watch for that PR to get merged (https://github.com/moby/moby/pull/45942)

**Comment by djaglowski** [comment]
> I'll watch for that PR to get merged ([moby/moby#45942](https://github.com/moby/moby/pull/45942))

Thanks @srabraham, it was just merged.

**Comment by srabraham** [comment]
Awesome, thanks! There's still a bit of an issue, since the latest testcontainers doesn't seem to be compatible with the latest moby/moby 😢 . I can wait though

```bash
git clone git@github.com:testcontainers/testcontainers-go.git
cd testcontainers-go
go build ./...  # works
go get -x github.com/docker/docker@dab9ffb252186f4c47416addb136d579f9314c6a
go build ./... 
go build ./...  # outputs the below:
# github.com/testcontainers/testcontainers-go
./container.go:67:36: undefined: types.AuthConfig
./container.go:78:34: undefined: types.AuthConfig
./container.go:207:62: undefined: types.AuthConfig
./container.go:210:27: undefined: types.AuthConfig
./container.go:213:34: undefined: types.AuthConfig
./docker_auth.go:17:72: undefined: types.AuthConfig
./docker_auth.go:23:26: undefined: types.AuthConfig
./docker_auth.go:30:25: undefined: types.AuthConfig
./docker_auth.go:53:47: undefined: types.AuthConfig
./docker_auth.go:59:27: undefined: types.AuthConfig
```

**Comment by srabraham** [comment]
Looks like Go might be doing a partial revert on account of this and related bugs, so we may just want to wait and avoid Go 1.20.6. https://github.com/golang/go/issues/61431

**Comment by smhmayboudi** [comment]
it would be very nice to check this issue.

**Comment by joshiste** [comment]
> As I mentioned in the original post, this issue is introduced in Go 1.20.6, via [golang/go#61076](https://github.com/golang/go/issues/61076)
> 
> @Slijkhuis: thanks! I'll watch for that PR to get merged ([moby/moby#45942](https://github.com/moby/moby/pull/45942))


Imho the PR to watch is https://github.com/moby/moby/pull/45977 which already got merged into the `23.0` branch - but is unreleased yet. 
I've tested testcontainers-go using the `23.0` branch and go v1.20.6 which worked for me.

So to get an immediate pain relief testcontainers-go users can add a replacement to their `go.mod´
```
replace github.com/docker/docker v23.0.5+incompatible => github.com/docker/docker v23.0.7-0.20230730020554-801e90549aac+incompatible
```

(until the docker team has released the 23.0.7 and the testcontainers-go team has updated)

**Comment by mdelapenya** [comment]
I think we have a potential fix here: https://github.com/testcontainers/testcontainers-go/pull/1394
